### ISSUE TITLE
Support nikic/PHP-Parser 5.x

### ISSUE THREAD

**Issue by staabm** [PHPCSStandards/PHP_CodeSniffer#357]
## Is your feature request related to a problem?

incompatibility of PHP-Parser 4.x and 5.x, see https://github.com/nikic/PHP-Parser/issues/981

## Describe the solution you'd like

PHP-CS ships with a bunch of polyfill values for Tokens in https://github.com/PHPCSStandards/PHP_CodeSniffer/blob/45c61f55b6ead1a86212c0a75c7a635456a2844e/src/Util/Tokens.php

PHP-Parser 5.x requires these token values to be integers. currently these polyfill tokens contain string values.

## Additional context (optional)

- [ ] I intend to create a pull request to implement this feature.

**Comment by jrfnl** [comment]
@staabm I'm aware of this, but I am not aware of a real-life use-case which would justify making breaking changes in PHPCS for something introduced at a later date in an unrelated tool, which should not be loaded while PHPCS is running anyway.

What I mean is:
* PHPCS does not load anything via the Composer autoloader which would cause this conflict to be a problem for other projects.
* Similarly, PHP-Parser should not be (auto-)loaded when running PHPCS, so again, this shouldn't cause a problem.

The [only real-life case](https://github.com/sebastianbergmann/php-code-coverage/issues/1025) I've seen/found so far is for the code coverage runs on PHPCS **itself** and only when running on PHP 8.0 with PHPUnit 9.x (9.3 or higher), which is a situation in which both projects would be loaded at the same time. And that has already been solved in #245.

**Comment by jrfnl** [comment]
P.S.: aside from the above, I suggest reporting this to PHP-Parser as they are the ones causing the problem.

**Comment by staabm** [comment]
> I suggest reporting this to PHP-Parser

its already reported, see https://github.com/nikic/PHP-Parser/issues/981

**Comment by jrfnl** [comment]
@staabm In that case, let's close this issue as reported to the wrong project.

**Comment by staabm** [comment]
IMO the cause of the problem is here - but its up to you

**Comment by jrfnl** [comment]
@staabm No, the cause of the problem is in PHP-Parser. PHPCS does not use PHP-Parser, so there is no reason for PHPCS to comply with whatever arbitrary (okay, not completely arbitrary, but still) demands PHP-Parser puts on these _global_ PHP constants, which PHPCS polyfills.

**Comment by jrfnl** [comment]
For the record: the root cause of the use case which was the trigger for this issue was an old version of `phpDocumenter/TypeResolver` being used and had nothing to do with PHPCS (at all).
